oauth2_http/java/com/google/auth/oauth2/IdentityPoolCredentials.java

```
@@ -0,0 +1,311 @@
+ */
+public class IdentityPoolCredentials extends ExternalAccountCredentials {
+
+  /**
+   * The IdentityPool credential source. Dictates the retrieval method of the external credential,
+   * which can either be through a metadata server or a local file.
+   */
+  static class IdentityPoolCredentialSource extends ExternalAccountCredentials.CredentialSource {
+
+    enum IdentityPoolCredentialSourceType {
+      FILE,
+      URL
+    }
+
+    enum CredentialFormatType {
+      TEXT,
+      JSON
+    }
+
+    private IdentityPoolCredentialSourceType credentialSourceType;
+    private CredentialFormatType credentialFormatType;
+    private String credentialLocation;
+
+    @Nullable private String subjectTokenFieldName;
+    @Nullable private Map<String, String> headers;
+
+    /**
+     * The source of the 3P credential.
+     *
+     * <p>If this is a file based 3P credential, the credentials file can be retrieved using the
+     * `file` key.
+     *
+     * <p>If this is URL-based 3p credential, the metadata server URL can be retrieved using the
+     * `url` key.
+     *
+     * <p>The third party credential can be provided in different formats, such as text or JSON. The
+     * format can be specified using the `format` header, which returns a map with keys `type` and
+     * `subject_token_field_name`. If the `type` is json, the `subject_token_field_name` must be
+     * provided. If no format is provided, we expect the token to be in the raw text format.
+     *
+     * <p>Optional headers can be present, and should be keyed by `headers`.
+     */
+    public IdentityPoolCredentialSource(Map<String, Object> credentialSourceMap) {
+      super(credentialSourceMap);
+
+      if (credentialSourceMap.containsKey("file")) {
+        credentialLocation = (String) credentialSourceMap.get("file");
+        credentialSourceType = IdentityPoolCredentialSourceType.FILE;
+      } else if (credentialSourceMap.containsKey("url")) {
```

What if it contains both?

---

you asked this last time - it will default to file

---

yes, but does that make sense? What does the caller want if they supply both? Is it an error that should throw an exception?

---

Added an exception